python/repeated.c

```diff
@@ -0,0 +1,716 @@
+  }
+  PyObject* ret = PyObject_RichCompare(list1, list2, opid);
+  Py_DECREF(list1);
+  Py_XDECREF(del);
+  return ret;
+}
+
+static PyObject* PyUpb_RepeatedContainer_Subscript(PyObject* _self,
+                                                   PyObject* key) {
+  PyUpb_RepeatedContainer* self = (PyUpb_RepeatedContainer*)_self;
+  upb_array* arr = PyUpb_RepeatedContainer_GetIfWritable(self);
+  Py_ssize_t size = arr ? upb_array_size(arr) : 0;
+  Py_ssize_t idx, count, step;
+  if (!IndexToRange(key, size, &idx, &count, &step)) return NULL;
+  const upb_fielddef* f = PyUpb_RepeatedContainer_GetField(self);
+  if (step == 0) {
+    return PyUpb_UpbToPy(upb_array_get(arr, idx), f, self->arena);
+  } else {
+    PyObject* ret = PyList_New(count);
+    for (Py_ssize_t i = 0; i < count; i++, idx += step) {
+      upb_msgval msgval = upb_array_get(self->ptr.arr, idx);
+      PyObject* item = PyUpb_UpbToPy(msgval, f, self->arena);
+      PyList_SetItem(ret, i, item);
+    }
+    return ret;
+  }
+}
+
+static int PyUpb_RepeatedContainer_SetSubscript(
+    PyUpb_RepeatedContainer* self, upb_array* arr, const upb_fielddef* f,
+    Py_ssize_t idx, Py_ssize_t count, Py_ssize_t step, PyObject* value) {
+  upb_arena* arena = PyUpb_Arena_Get(self->arena);
+  if (upb_fielddef_issubmsg(f)) {
+    PyErr_SetString(PyExc_TypeError, "does not support assignment");
+    return -1;
+  }
+
+  if (step == 0) {
+    // Set single value.
+    upb_msgval msgval;
+    if (!PyUpb_PyToUpb(value, f, &msgval, arena)) return -1;
+    upb_array_set(arr, idx, msgval);
+    return 0;
+  }
+
+  // Set range.
+  PyObject* seq =
+      PySequence_Fast(value, "must assign iterable to extended slice");
+  PyObject* item = NULL;
+  int ret = -1;
+  if (!seq) goto err;
+  if (PySequence_Size(seq) != count) {
+    PyErr_Format(PyExc_ValueError,
+                  "attempt to assign sequence of size %zd to extended slice "
+                  "of size %zd",
+                  PySequence_Size(seq), count);
+    goto err;
+  }
+  for (Py_ssize_t i = 0; i < count; i++, idx += step) {
+    upb_msgval msgval;
+    item = PySequence_GetItem(seq, i);
+    if (!item) goto err;
+    // XXX: if this fails we can leave the list partially mutated.
+    if (!PyUpb_PyToUpb(item, f, &msgval, arena)) goto err;
+    Py_DECREF(item);
+    item = NULL;
+    upb_array_set(arr, idx, msgval);
+  }
+  ret = 0;
+
+err:
+  Py_XDECREF(seq);
+  Py_XDECREF(item);
+  return ret;
+}
+
+static int PyUpb_RepeatedContainer_DeleteSubscript(upb_array* arr,
+                                                   Py_ssize_t idx,
+                                                   Py_ssize_t count,
+                                                   Py_ssize_t step) {
+  // Normalize direction: deletion is order-independent.
+  Py_ssize_t start = idx;
+  if (step < 0) {
+    Py_ssize_t end = start + step * (count - 1);
+    start = end;
+    step = -step;
+  }
+  size_t dst = start;
+  size_t src = step ? start + 1 : start + count;
```

Comment:
I think this is not right, but I am having a bit of trouble fitting it in my head precisely (so I could easily be wrong here).

Consider
```
idx = 0
count = 4
step = 3
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
```

I think you want the call sequence

```
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
move(0, 1, 2);
[1, 2, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
move(2, 4, 2);
[1, 2, 4, 5, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
move(4, 7, 2);
[1, 2, 4, 5, 7, 8, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
move(6, 10, 2);
[1, 2, 4, 5, 7, 8, 10, 11, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
move(8, 12, 8);
[1, 2, 4, 5, 7, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 16, 17, 18, 19]
resize(16);
[1, 2, 4, 5, 7, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
```

I think this goes wrong right from the very first `src = start + count` when it should always start at `src = start + 1`.

I think unit tests what we really need for this one.

Reply:
Done (unit tests are available now, and the bug is fixed).